### ISSUE TITLE
When using Ray, bitsandbytes says "compiled without GPU support"

### ISSUE THREAD

**Issue by drikster80** [axolotl-ai-cloud/axolotl#2869]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

**Background:** Running in custom container on Kubernetes distributed across 2 nodes. Validated bitsandbytes was compiled with GPU support and validated simple 8-bit quantization on a model uses GPU.

**Package Versions:**
Package                                  Version
---------------------------------------- -------------------
axolotl                                      0.10.1
bitsandbytes                            0.46.1
flash-attn                                 2.8.0.post2
ray                                           2.47.1

When I run a simple test on a single node with the 8b-qat-fsdp2.yml example (only removing the qat configs and changing to 14b):
`axolotl train examples/qwen3/14b-fsdp2.yml.yml`

No BnB warnings are present.

### Current behaviour

When launching across the 2-node ray cluster, same unmodified container, via `axolotl train examples/qwen3/1.7b-fsdp2.yml.yml --use-ray`, I receive the error:
```
(TrainTrainable pid=1735, ip=192.168.192.13) [2025-07-04 13:37:09,303] [WARNING] [bitsandbytes.cextension.get_native_library:282] [PID:1735] The installed version of bitsandbytes was compiled without GPU support. 8-bit optimizers and GPU quantization are unavailable.
```

### Steps to reproduce

1. Update examples/qwen3/8b-qat-fsdp2.yml (copied to 'examples/qwen3/1.7b-fsdp2.yml')
	- Change to 1.7B
	- Remove QAT configs
	- Remove FSDP configs

2. Launch on single node (no Warnings):
	- `axolotl train examples/qwen3/1.7b-fsdp2.yml`
	
3. Update config to support Ray Cluster
	- Update to add
```yaml
use_ray: true
ray_num_workers: 2
resources_per_worker:
    GPU: 1
```

4. Launch with Ray (Warning about bitsandbytes:
	- `axolotl train examples/qwen3/1.7b-fsdp2.yml --use-ray`

### Config yaml

```yaml
base_model: /models/Qwen3-1.7B
# Automatically upload checkpoint and final model to HF
# hub_model_id: username/custom_model_name

load_in_8bit: false
load_in_4bit: false
strict: false

plugins:
  - axolotl.integrations.liger.LigerPlugin

liger_rope: true
liger_rms_norm: true
liger_glu_activation: true
liger_layer_norm: true
liger_fused_linear_cross_entropy: true

datasets:
  - path: tatsu-lab/alpaca
    type: alpaca

output_dir: /models/Qwen3-1.7B/outputs/qat_out/

sequence_len: 2048
sample_packing: true
flex_attention: true
pad_to_sequence_len: true

flex_attn_compile_kwargs:
  dynamic: false
  mode: max-autotune-no-cudagraphs

wandb_project:
wandb_entity:
wandb_watch:
wandb_name:
wandb_log_model:

gradient_accumulation_steps: 1
micro_batch_size: 2
max_steps: 2000
optimizer: adamw_torch_fused
lr_scheduler: cosine
learning_rate: 2e-5

bf16: true
tf32: true

resume_from_checkpoint:
logging_steps: 1

evals_per_epoch: 1
saves_per_epoch: 1

warmup_steps: 10
weight_decay: 0.0
fsdp:
  - full_shard
  - auto_wrap

fsdp_config:
  fsdp_version: 2
  fsdp_offload_params: false
  fsdp_cpu_ram_efficient_loading: true
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_transformer_layer_cls_to_wrap: Qwen3DecoderLayer
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_sharding_strategy: FULL_SHARD
  fsdp_reshard_after_forward: true
  fsdp_activation_checkpointing: true

special_tokens:

use_ray: true
ray_num_workers: 2
resources_per_worker:
    GPU: 1
```

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12.11

### axolotl branch-commit

release/v0.10.1

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by drikster80** [comment]
One additional note:
When running `python3 -m bitsandbytes` within the container, it properly shows it can import and call the CUDA library.
```
=================== bitsandbytes v0.46.1 ===================
Platform: Linux-6.8.0-1028-nvidia-64k-aarch64-with-glibc2.35
  libc: glibc-2.35
Python: 3.12.11
PyTorch: 2.7.0+cu128
  CUDA: 12.8
  HIP: N/A
  XPU: N/A
Related packages:
  accelerate: 1.7.0
  diffusers: not found
  numpy: 2.0.1
  pip: 25.1.1
  peft: 0.15.2
  safetensors: 0.5.3
  transformers: 4.52.4
  triton: 3.3.0
  trl: 0.18.2
============================================================
PyTorch settings found: CUDA_VERSION=128, Highest Compute Capability: (9, 0).
Checking that the library is importable and CUDA is callable...
SUCCESS!
```

**Comment by winglian** [comment]
@kouroshHakha have you seen this before?

**Comment by winglian** [comment]
@drikster80 any chance this could be related to the user it ends up running under? https://github.com/bitsandbytes-foundation/bitsandbytes/issues/112#issuecomment-1807755284 

For example does Ray start a new process with a different python/conda env?

**Comment by drikster80** [comment]
> [@drikster80](https://github.com/drikster80) any chance this could be related to the user it ends up running under? [bitsandbytes-foundation/bitsandbytes#112 (comment)](https://github.com/bitsandbytes-foundation/bitsandbytes/issues/112#issuecomment-1807755284)
> 
> For example does Ray start a new process with a different python/conda env?

AFAIK, everything is as root. I'm using the [VLLM Ray example script](https://github.com/vllm-project/vllm/blob/main/examples/online_serving/multi-node-serving.sh) and LeaderWorkerSet to start the Ray cluster. 

Thinking about this, I wonder if some environment variable isn't getting distributed with ray (e.g. LD_LIBRARY_PATH or CUDA_HOME or something?) that would allow bitsandbytes to pickup the CUDA library. I'll do some testing directly with ray to better isolate the problemspace.

**Comment by drikster80** [comment]
Okay, I found the problem. It's not with Axolotl, so closing the ticket and will dig deeper to see if it's a problem with Ray (likely) or the library.

When running in ray, the environment variable (CUDA_VISIBLE_DEVICES="") is created and empty and either bitsandbytes or pytorch look for whether that envar exists. If it does exist, it seems like bitsandbytes fails to get the pytorch CUDA config or something. By simply removing that envar from the Ray environment, bitsandbytes is able to successfully use CUDA. 

```python
   try:
        # Fix CUDA_VISIBLE_DEVICES if it's empty (this hides GPUs from PyTorch)
        env = os.environ.copy()
        cuda_visible_devices = env.get("CUDA_VISIBLE_DEVICES", "")
        
        if cuda_visible_devices == "":
            # Remove CUDA_VISIBLE_DEVICES entirely to let PyTorch see all GPUs
            if "CUDA_VISIBLE_DEVICES" in env:
                del env["CUDA_VISIBLE_DEVICES"]
            result["cuda_devices_fix"] = "Removed empty CUDA_VISIBLE_DEVICES"
        else:
            result["cuda_devices_fix"] = f"CUDA_VISIBLE_DEVICES was: '{cuda_visible_devices}'"
```

**Comment by drikster80** [comment]
Re-opened the issue. My isolated ray test script had a bug, and Ray did work as expected in a separate environment, so still troubleshooting this one.

@kouroshHakha, any thoughts on this? I'm not as familiar with how Ray launched the remote environment and still digging into it.
> https://github.com/bitsandbytes-foundation/bitsandbytes/issues/112#issuecomment-1807755284
> 
> For example does Ray start a new process with a different python/conda env?

**Comment by drikster80** [comment]
Looks like the raylet log file `/tmp/ray/session_latest/logs/raylet.out` shows the culprit.

```
[state-dump] Resource usage {
[state-dump]     - (language=PYTHON actor_or_task=validate_bitsandbytes_cuda pid=576 worker_id=83d90062c2968f842dd411d205bc15b83326821af8baa3644c9456a3): {CPU: 1, node:192.168.165.177: 0.001}
[state-dump]     - (language=PYTHON actor_or_task=validate_bitsandbytes_cuda pid=574 worker_id=33c882a3914cee0a02cfa27fe39d06b6a58babcad908a94fac18d5ee): {node:__internal_head__: 0.001, CPU: 1}
```

When the `validate_bitsandbytes_cuda()` is scheduled, it only passes in `CPU: 1` and no GPUs. Not sure where that is coming from though. Still digging.

**Comment by drikster80** [comment]
It looks like that log might be a false-positive while the job is getting setup, but not actually effect the job.

When I set envvar `BNB_CUDA_VERSION=128`, I can see bnb gets initialized multiple times in the logs.

**Pre-Ray initialization:**
```
[2025-07-07 14:12:20,690] [WARNING] [bitsandbytes.cextension.get_cuda_bnb_library_path:31] [PID:4796] WARNING: BNB_CUDA_VERSION=128 environment variable detected; loading libbitsandbytes_cuda128.so.
```

**Potential false-positive log**
```
(TrainTrainable pid=5054) [2025-07-07 14:12:29,409] [WARNING] [bitsandbytes.cextension.get_native_library:282] [PID:5054] The installed version of bitsandbytes was compiled without GPU support. 8-bit optimizers and GPU quantization are unavailable.
```

**BNB Called again after training started with no issues:**
```
(RayTrainWorker pid=5157) [2025-07-07 14:12:35,775] [WARNING] [bitsandbytes.cextension.get_cuda_bnb_library_path:31] [PID:5157] WARNING: BNB_CUDA_VERSION=128 environment variable detected; loading libbitsandbytes_cuda128.so.
(RayTrainWorker pid=5157) This can be used to load a bitsandbytes version built with a CUDA version that is different from the PyTorch CUDA version.
(RayTrainWorker pid=5157) If this was unintended set the BNB_CUDA_VERSION variable to an empty string: export BNB_CUDA_VERSION=
(RayTrainWorker pid=5157)
(RayTrainWorker pid=611, ip=192.168.192.9)
(RayTrainWorker pid=611, ip=192.168.192.9) [2025-07-07 14:12:40,833] [WARNING] [bitsandbytes.cextension.get_cuda_bnb_library_path:31] [PID:611] WARNING: BNB_CUDA_VERSION=128 environment variable detected; loading libbitsandbytes_cuda128.so.
(RayTrainWorker pid=611, ip=192.168.192.9) This can be used to load a bitsandbytes version built with a CUDA version that is different from the PyTorch CUDA version.
(RayTrainWorker pid=611, ip=192.168.192.9) If this was unintended set the BNB_CUDA_VERSION variable to an empty string: export BNB_CUDA_VERSION=
```

**Comment by kouroshHakha** [comment]
@drikster80 The only thing ray might have done is depending on `num_gpus` set on the actor it will override the `CUDA_VISIBLE_DEVICES`. since you have `resources_per_worker:  GPU: 1` I don't think that it is setting it to empty strings. Could you print all the env vars within the train function can compare to the case that works?


To debug this, you could also try isolating the problem to a simple actor:

```python

@ray.remote(num_gpus=1)
class Actor:

      def __init__(self):
             # run the same code that initialized b&b
      
      def run(self):
            print("hello")



actor = Actor.remote()
ray.get(actor.run.remote())
```
